avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java

```diff
@@ -71,4 +78,35 @@ protected Object deserialize(byte[] payload) throws SerializationException {
       throw new SerializationException("Error deserializing Avro message for id " + id, e);
     }
   }
+
+  private DatumReader getDatumReader(Schema writerSchema) {
+    if (useSpecificAvroReader) {
+      return new SpecificDatumReader(writerSchema, getReaderSchema(writerSchema));
+    } else {
+      return new GenericDatumReader(writerSchema);
+    }
+  }
+
+  private Schema getReaderSchema(Schema writerSchema) {
+    Schema readerSchema = targetSchemaCache.get(writerSchema);
+    if (readerSchema == null) {
+      Class targetClass = SpecificData.get().getClass(writerSchema);
```

Comment:
One thing I'm not clear on is how this handles different versions of the same schema. And the javadoc for this method doesn't give any details either. For example, if I have version 1 and version 2 of the schema registered, does getClass() only look at the `namespace` and `name` fields to look up the class, or does it actually compare the schema as well.

I think we need to think through the use cases to decide what the right behavior is. For example, if Avro only uses the name, should we also compare schemas and trigger an exception if they are incompatible? Or just leave that to the deserialization step which would presumably also fail? Would it be more useful to catch that early since we might be able to give a more meaningful error message.